pyqtgraph/parametertree/parameterTypes/basetypes.py

```diff
@@ -0,0 +1,419 @@
+        Called when the widget's value is changing, but not finalized.
+        For example: editing text before pressing enter or changing focus.
+        """
+        self.param.sigValueChanging.emit(self.param, self.widget.value())
+
+    def selected(self, sel):
+        """Called when this item has been selected (sel=True) OR deselected (sel=False)"""
+        ParameterItem.selected(self, sel)
+
+        if self.widget is None:
+            return
+        if sel and self.param.writable():
+            self.showEditor()
+        elif self.hideWidget:
+            self.hideEditor()
+
+    def showEditor(self):
+        self.widget.show()
+        self.displayLabel.hide()
+        self.widget.setFocus(QtCore.Qt.FocusReason.OtherFocusReason)
+
+    def hideEditor(self):
+        self.widget.hide()
+        self.displayLabel.show()
+
+    def limitsChanged(self, param, limits):
+        """Called when the parameter's limits have changed"""
+        ParameterItem.limitsChanged(self, param, limits)
+
+    def defaultChanged(self, param, value):
+        self.updateDefaultBtn()
+
+    def treeWidgetChanged(self):
+        """Called when this item is added or removed from a tree."""
+        ParameterItem.treeWidgetChanged(self)
+
+        ## add all widgets for this item into the tree
+        if self.widget is not None:
+            tree = self.treeWidget()
+            if tree is None:
+                return
+            if self.asSubItem:
+                self.subItem.setFirstColumnSpanned(True)
+                tree.setItemWidget(self.subItem, 0, self.widget)
+            tree.setItemWidget(self, 1, self.layoutWidget)
+            self.displayLabel.hide()
+            self.selected(False)
+
+    def defaultClicked(self):
+        self.param.setToDefault()
+
+    def optsChanged(self, param, opts):
+        """Called when any options are changed that are not
+        name, value, default, or limits"""
+        ParameterItem.optsChanged(self, param, opts)
+
+        if 'enabled' in opts:
+            self.updateDefaultBtn()
+            self.widget.setEnabled(opts['enabled'])
+
+        if 'readonly' in opts:
+            self.updateDefaultBtn()
+            if hasattr(self.widget, 'setReadOnly'):
+                self.widget.setReadOnly(opts['readonly'])
+            else:
+                self.widget.setEnabled(self.param.opts['enabled'] and not opts['readonly'])
+
+        if 'tip' in opts:
+            self.widget.setToolTip(opts['tip'])
+
+
+class EventProxy(QtCore.QObject):
+    def __init__(self, qobj, callback):
+        QtCore.QObject.__init__(self)
+        self.callback = callback
+        qobj.installEventFilter(self)
+
+    def eventFilter(self, obj, ev):
+        return self.callback(obj, ev)
+
+
+class SimpleParameter(Parameter):
+    """Parameter representing a single value.
+
+    This parameter is backed by :class:`WidgetParameterItem` to represent the
+    following parameter names through various subclasses:
+      - 'int'
```

Comment:
I think this might want a blank line here, before the bullet list?

Reply:
Looks like that was the one -- thanks :)